File: healthy_food_bank_flutter/lib/models/product_demand_summary.dart
/// Model for aggregated product demand summary
/// Used to show total quantity needed per product for vendor inventory planning
class ProductDemandSummary {
  final String productName;
  final int? productId;
  final int totalQuantity;
  final String? unit;

  ProductDemandSummary({
    required this.productName,
    this.productId,
    required this.totalQuantity,
    this.unit,
  });

  factory ProductDemandSummary.fromJson(Map<String, dynamic> json) {
    return ProductDemandSummary(
      productName: json['productName'] as String,
      productId: json['productId'] as int?,
      totalQuantity: json['totalQuantity'] as int,
      unit: json['unit'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'productName': productName,
      'productId': productId,
      'totalQuantity': totalQuantity,
      'unit': unit,
    };
  }

  @override
  String toString() {
    return 'ProductDemandSummary(productName: $productName, totalQuantity: $totalQuantity, unit: $unit)';
  }
}

/// Response model for vendor order summary
class VendorOrderSummaryResponse {
  final List<ProductDemandSummary> products;
  final int totalOrders;
  final int totalProducts;

  VendorOrderSummaryResponse({
    required this.products,
    required this.totalOrders,
    required this.totalProducts,
  });

  factory VendorOrderSummaryResponse.fromJson(Map<String, dynamic> json) {
    return VendorOrderSummaryResponse(
      products: (json['products'] as List<dynamic>)
          .map((item) => ProductDemandSummary.fromJson(item as Map<String, dynamic>))
          .toList(),
      totalOrders: json['totalOrders'] as int,
      totalProducts: json['totalProducts'] as int,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'products': products.map((p) => p.toJson()).toList(),
      'totalOrders': totalOrders,
      'totalProducts': totalProducts,
    };
  }

  @override
  String toString() {
    return 'VendorOrderSummaryResponse(totalProducts: $totalProducts, totalOrders: $totalOrders)';
  }
}

File: order-service/src/main/java/order_service/dto/ProductDemandSummary.java
package order_service.dto;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * DTO for aggregated product demand summary
 * Used to show total quantity needed per product for vendor inventory planning
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
public class ProductDemandSummary {
    private String productName;
    private Long productId;
    private Integer totalQuantity;
    private String unit;

    // Constructor for SQL projection (without productId and unit)
    public ProductDemandSummary(String productName, Integer totalQuantity) {
        this.productName = productName;
        this.totalQuantity = totalQuantity;
    }

    // Constructor for SQL projection (with unit) - Integer version
    public ProductDemandSummary(String productName, Integer totalQuantity, String unit) {
        this.productName = productName;
        this.totalQuantity = totalQuantity;
        this.unit = unit;
    }

    // Constructor for SQL projection (with unit) - Double version
    // Used when SUM() returns Double from database
    public ProductDemandSummary(String productName, Double totalQuantity, String unit) {
        this.productName = productName;
        this.totalQuantity = totalQuantity != null ? totalQuantity.intValue() : 0;
        this.unit = unit;
    }
}
